Routing windowslivewriter manifest & XMLRPC MetaWeblog to support OpenLiveWriter

warrenbuckley opened the pull request:
* Adds the `wlwmanifest/{id}` route to return the XML used in conjunction with the RSD XML
* Gets the XMLRPC / MetaWeblog Controller for receiving HTTP POSTed XML for OpenLiveWriter
* Drops Cookcomputing.XMLRpc for `WilderMinds.MetaWeblog`
https://github.com/shawnwildermuth/MetaWeblog

## Help needed
I am unsure with DI to update the Root Articulate Blog Node ID - so appreciate any pointers @Shazwazza 
```csharp
private IPublishedContent BlogRoot()
{
    // TODO: Do not hardcode 1068 - need to pass it through from the controller route
    // When injecting this service into the controller - but how?!?
    var node = _umbracoContextAccessor.GetRequiredUmbracoContext().Content.GetById(1068);
    ...
}
```

## Testing
* Download Open Live Writer
* Have test site running & put a breakpoint in `MetaWeblogController` to help with ddebugging
* Add new configuration pointing to example site
* Add a new post - via theme detection
* It should delete the temp post & put it in the recycle bin
* Add a new post
  * Select existing category
  * Add a new category
  * Select existing tags
  * Add a new tag
  * Add one or more image (Use the silly effects in LiveWriter)
  * Save draft & verify
  * Publish & verify
* Close Open Live Writer
* File -> Open Recent Post -> Select blog account from list (Fetches recent posts)
  * Select an item
  * Use delete button - ensure blog post is moved to recycle bin
  * Select an item to edit
  * Update content, properties, categories & tags - verify changes are expected

Commit:
Add in Windows Live Writer Manifest route

Commit:
Converts from PluginController to RenderController for ArticulateRouter to work with it

Shazwazza commented:
Has some conflicts now. Then windows live writer integration will need to be tested

Commit:
Merge remote-tracking branch 'upstream/release/5.0.0' into routing/windowslivewriter-manifest

# Conflicts:
#	src/Articulate/Routing/ArticulateRouter.cs

Commit:
Move down in hope to make next merge conflict a bit easier

warrenbuckley commented:
This is just the route and XML returned for the WLW Manifest
![image](https://user-images.githubusercontent.com/1389894/177521430-04e8d9a4-a3f8-44d8-9bec-f2841a983ae7.png)

Will do a sperate PR for the route that is needed for WLW (Will also see if you can download/use it)

warrenbuckley commented:
OK not sure if Windows Live Writer is a thing anymore and the MetaWebBlog Spec is 404 from your comments
http://xmlrpc.com/metaWeblogApi.html

Is this something still worth supporting and migrating over @Shazwazza ?

Shazwazza commented:
It's a thing, It's Open Live Writer now https://openlivewriter.com/

warrenbuckley commented:
Ok will have to take a look to try and figure out the rest of it.

Gotta love the XP/Windows Vista style vibes of the app still !

Commit:
Merge remote-tracking branch 'origin/release/5.0.0' into routing/windowslivewriter-manifest

# Conflicts:
#	src/Articulate/Routing/ArticulateRouter.cs

Commit:
Update MetaWeblogRouteContstraint aka OpenLiveWriter stuff

warrenbuckley commented:
Making draft PR as will include the other routing stuff to get all of the MetaWeblog/Open Live Writer stuff working

Commit:
Uncomment class as I believe we need this

Commit:
Uncomment old MetaWeblog Handler class and fix up errors, references and DI bits

Commit:
Looking at how it was previously routed but not sure how to wire this up

warrenbuckley commented:
OK @Shazwazza I am gonna need your help here as I have fixed up the MetaWeblogHandler class that was commented out but I have no idea how to wire it up into the routes with RouteTransformer stuff you have.

Any advice or pointers you can give on how to fix/resolve this then I can try and attempt it

![No Idea](https://media.giphy.com/media/3o6UBil4zn1Tt03PI4/giphy.gif)

Commit:
Merge branch 'release/5.0.0' into routing/windowslivewriter-manifest

Commit:
Merge branch 'release/5.0.0' into routing/windowslivewriter-manifest

Commit:
WIP to get MetaWeblog XMLRPC support working for OpenLiveWriter

Using WilderMinds.MetaWeblog package to help with this - may compare to other solutions

Commit:
From composer remove reference to MetaWeblog Route handler & factory

Commit:
Port code from MetaWeblogHandler code

Commit:
Bug but unsure why the richText property is not returning HtmlString - so fixing so that it does

Commit:
RTE in newer Umbraco return IHtmlEncodedString and not HtmlString

Updated MarkdownPropertyEditor Value Convertor to return the same type IHtmlEncodedString

warrenbuckley commented:
@Shazwazza ready for review & testing with Open Live Writer now :)

![Ready when you are](https://media.giphy.com/media/vH2LoAZUzqLenXtuSG/giphy.gif)

Commit:
Remove unused models for MetaWeblog

Commit:
Comment out the handler code - just so we still have a reference

We can delete this once PR merged in and all good

Commit:
Remove the port of CookComputing.XMLRPC package known as Kveer.XmlRPC & updat excetions to be more generic

Commit:
Attempt to get the Blog Root ID passed into the MetaWeblogProvider - but still no luck

Shazwazza commented on src/Articulate/Controllers/MetaWeblogController.cs:
```diff
@@ -0,0 +1,49 @@
+        public MetaWeblogController(ILogger<RenderController> logger,
+            ICompositeViewEngine compositeViewEngine,
+            IUmbracoContextAccessor umbracoContextAccessor,
+            ArticulateMetaWeblogProvider articulateMetaWeblogProvider,
```
Shouldn't IMetaWeblogProvider be injected here? ArticulateMetaWeblogProvider is the implementation

Shazwazza replied:
Resolving this comment since this has been implemented/fixed in a different way. We now manually create the ArticulateMetaWeblogProvider based on the current node and then manually create the MetaWeblogService to process the xmlrpc call.

Shazwazza commented on src/Articulate/Controllers/MetaWeblogController.cs:
```diff
@@ -0,0 +1,49 @@
+            }
+
+            // Set the Blog Root Node ID in the implementation
+            _articulateMetaWeblogProvider.ArticulateBlogRootNodeId = id;
```
Can you verify that this id is not zero?

Shazwazza commented on src/Articulate/Components/ArticulateComposer.cs:
```diff
@@ -66,9 +56,13 @@ public override void Compose(IUmbracoBuilder builder)
+            //services.AddMetaWeblog<ArticulateMetaWeblogProvider>();
+            services.AddScoped<ArticulateMetaWeblogProvider>();
+            services.AddScoped<IMetaWeblogProvider, ArticulateMetaWeblogProvider>();
+            services.AddScoped<MetaWeblogService>();
```
Where is this service coming from?

Shazwazza commented on src/Articulate/Components/ArticulateComposer.cs:
```diff
@@ -66,9 +56,13 @@ public override void Compose(IUmbracoBuilder builder)
 
+            // MetaWeblog - https://github.com/shawnwildermuth/MetaWeblog
+            //services.AddMetaWeblog<ArticulateMetaWeblogProvider>();
+            services.AddScoped<ArticulateMetaWeblogProvider>();
```
Is there a requirement to register both the concrete and abstraction here?

Shazwazza commented on src/Articulate/Controllers/MetaWeblogController.cs:
```diff
@@ -0,0 +1,49 @@
+            _articulateMetaWeblogProvider.ArticulateBlogRootNodeId = id;
+
+            // This service call from the Nuget package - consumes our implmentation/service of Metaweblog above
+            string result = await _metaWeblogService.InvokeAsync(rawContent);
```
From what I can tell the articulate root node is always zero when your provider is executed because:

* You have registered both the concrete ArticulateMetaWeblogProvider and the abstraction of IMetaWeblogProvider as separate services
* Assuming that IMetaWeblogProvider is injected into this MetaWeblogService - which will be a different instance of the ArticulateMetaWeblogProvider instance you are modifying here

To fix that you need to change the DI registration to be a factory for IMetaWeblogProvider. So you would register the concrete ArticulateMetaWeblogProvider as you are doing, and then when registering the IMetaWeblogProvider you'd use `services.AddScoped<IMetaWeblogProvider>(s => s.GetRequiredService<ArticulateMetaWeblogProvider>());` which means that for both ArticulateMetaWeblogProvider and IMetaWeblogProvider you'd get the same instance.

Shazwazza commented on src/Articulate/MetaWeblog/ArticulateMetaWeblogProvider.cs:
```diff
@@ -0,0 +1,512 @@
+                content.SetInvariantOrDefaultCultureValue("excerpt", post.mt_excerpt, contentType, _localizationService);
+            }
+
+            // TODO: Investigate why this not supported ?!
```
Was this an existing TODO or a new one? If it's new, do we have a new task for this?

Shazwazza replied:
Comments will be supported once this PR is merged https://github.com/shawnwildermuth/MetaWeblog/pull/16

Shazwazza commented on src/Articulate/MetaWeblog/ArticulateMetaWeblogProvider.cs:
```diff
@@ -0,0 +1,512 @@
+        private IPublishedContent BlogRoot()
+        {
+            // Why is ArticulateBlogRootNodeId always 0 & not the value set in the Controller ?
+            // TODO: Need Shan help to fix this hardcoded 1068
```
Have made another comment on how to fix this

Shazwazza commented on src/Articulate/MetaWeblog/ArticulateMetaWeblogProvider.cs:
```diff
@@ -0,0 +1,512 @@
+            mt_keywords = string.Join(",", post.Tags.ToArray()),
+            title = post.Name
+
+            // TODO: AllowComments & Author
```
new TODO or old TODO?

Shazwazza replied:
Will resolve, comments will be fixed when the https://github.com/shawnwildermuth/MetaWeblog/pull/16 is merged, we don'tneed to worry about author.

Shazwazza commented on src/Articulate/Routing/MetaWeblogHandler.cs:
```diff
@@ -7,43 +7,75 @@
 //using Articulate.Models.MetaWeblog;
 //using CookComputing.XmlRpc;
 //using HeyRed.MarkdownSharp;
+//using Microsoft.AspNetCore.Http;
```
Guess this file can be removed?

Shazwazza committed:
Merge branch 'release/5.0.0' into routing/windowslivewriter-manifest

# Conflicts:
#	src/Articulate/Articulate.csproj
#	src/Articulate/Models/PostModel.cs
#	src/Articulate/Routing/ArticulateRouter.cs
#	src/Articulate/Routing/MetaWeblogHandler.cs

Shazwazza committed:
Fixes metaweblog to work with any articulate root node.

Shazwazza committed:
remove file

Shazwazza committed:
adds validation

Shazwazza committed:
updates pacakges removes todos

Shazwazza commented:
Thanks warren, i fixed a bunch up in your branch and merged 🎉
